rclcpp_components/src/node_main.cpp.in

```diff
@@ -0,0 +1,66 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "class_loader/class_loader.hpp"
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp_components/node_factory.hpp"
+#include "rclcpp_components/node_factory_template.hpp"
+
+#define NODE_MAIN_LOGGER_NAME "@node@"
+
+int main(int argc, char * argv[])
+{
+  auto args = rclcpp::init_and_remove_ros_arguments(argc, argv);
+  rclcpp::Logger logger = rclcpp::get_logger(NODE_MAIN_LOGGER_NAME);
+  rclcpp::executors::SingleThreadedExecutor exec;
+  rclcpp::NodeOptions options;
+  options.arguments(args);
```

Comment:
@skucheria @jacobperron why do we _remove_ ROS arguments and pass non-ROS arguments to the `NodeOptions`? I don't follow.

Reply:
The motivation was some of our demos take non-ROS arguments via the command-line. When we went turn them into components, it wasn't obvious how to pass along these non-ROS arguments to the code.
What's the recommended way to get non-ROS arguments?

Reply:
For example, image_tools parses options by passing the nodes arguments to this function:

https://github.com/ros2/demos/blob/97651912ebad439a6bc62c34671cad6a5bfee785/image_tools/src/options.cpp#L46

IMO, the demos should be refactored to make use of ROS parameters instead of custom CLI arguments (but that's beside the point).

Reply:
So, we had a discussion about this on https://github.com/ros2/rclcpp/pull/816 and https://github.com/ros2/rclpy/pull/405#discussion_r319069482. As a result, all `NodeOptions::arguments()` are implicitly ROS arguments. To achieve what you want here, you'd have to do `-- non-ros-arg0 ... non-ros-argN`, which is inconvenient but we were operating under the assumption that passing non-ROS arguments to a ROS node was a rare use case.

It looks like it's not @wjwwood @ivanpauno. Unless either of you strongly disagrees, I'll force `--ros-args`  to be given explicitly in both `rclpy` and `rclcpp`.

Reply:
:+1: to explicit

Reply:
> I'll force --ros-args to be given explicitly in both rclpy and rclcpp.

I think that's reasonable. I had forgotten this case.

Reply:
Is it then possible for `--ros-args` to clash with non-ros args? E.g. what if my node expects a CLI argument named `-p`?

Does `NodeOptions.arguments()` potentially return something like

`"--ros-args -p foo:=bar -- -p my_non_ros_arg"` ?

Reply:
Currently, `--ros-args` was prepended except if it was already part of the arguments.
@hidmic  sent a bunch of PRs for changing that, so `--ros-args` won't be prepended more.
`NodeOptions.arguments({"-p"})` will set a user specific argument.
`NodeOptions.arguments({"--ros-args", "-p"})` will set a ros specific argument.

IDK if that's exactly what you asked or not.

Reply:
@jacobperron even with the changes I'm making, that could be the case, yes. But here you're removing ROS arguments and thus `--ros-args ... --` sets will be removed. That also means that no `--ros-args` will reach your node as local arguments (they may still make it as global arguments) with this implementation, but it doesn't look like it cares about it.

Reply:
Thanks for the explanation.